src/solver/solve.jl

```diff
@@ -6,23 +6,78 @@ struct EfficiencyStats{TS <: Tuple, WT}
     walltime::WT
 end
 
+mutable struct GracefulExit
+    path::String
+    function GracefulExit(; path)
+        mkpath(path)
+        file = joinpath(path, "graceful_exit.dat")
+        open(file, "w") do io
+            println(io, "#      To gracefully exit job associated")
+            println(io, "#      with the output path:")
+            println(io, "#           $path")
+            println(io, "#      please change the following number from")
+            println(io, "#      0 (continue running) to 1 (gracefully exit)")
+            println(io, "0")
+        end
+        return new(path)
+    end
+end
+
 simulated_years_per_day(es::EfficiencyStats) =
     simulated_years(es) / walltime_in_days(es)
 
 simulated_years(es::EfficiencyStats) =
     (es.tspan[2] - es.tspan[1]) * (1 / (365 * 24 * 3600)) #=seconds * years per second=#
 walltime_in_days(es::EfficiencyStats) = es.walltime * (1 / (24 * 3600)) #=seconds * days per second=#
 
-function timed_solve!(integrator)
+function timed_solve!(integrator, graceful_exit::GracefulExit)
     walltime = @elapsed begin
         s = @timed_str begin
-            sol = SciMLBase.solve!(integrator)
+            _solve!(integrator, graceful_exit)
         end
     end
     @info "solve!: $s"
     es = EfficiencyStats(integrator.sol.prob.tspan, walltime)
     @info "sypd: $(simulated_years_per_day(es))"
-    return (sol, walltime)
+    return (integrator.sol, walltime)
+end
+
+finalize!(integrator) = DiffEqBase.finalize!(
+    integrator.callback,
+    integrator.u,
+    integrator.t,
+    integrator,
+)
+
+function exit_step_loop(ge::GracefulExit)
+    lines = readlines(joinpath(ge.path, "graceful_exit.dat"))
+    exit_loop = 0
+    for line in lines
+        startswith(line, "#") && continue
+        try
+            exit_loop = parse(Int, line)
+        catch
+            # Let's re-set the file, in
+            # case it was incorrectly edited.
+            GracefulExit(; path = ge.path)
+        end
+        break
+    end
+    return exit_loop == 1
+end
```

Comment:
Since this function is executed at the end of each step, I would suggest keeping it as simple as possible. 

Also, writing ascii files on a networked file system from different processes might lead to problems. 

I would recommend for this function to just read the first character, if it is not 0, kill the simulation gracefully. 

We should also consider the performance of opening and reading one file at each timestep on slow filesystems.

Reply:
That’s what I previously had but anticipated that people would want this somewhat documented, hence the comments